crates/ra_ide_api/src/name_ref_kind.rs

```diff
@@ -0,0 +1,99 @@
+use ra_db::FileId;
+use ra_syntax::{AstNode, AstPtr, ast};
+use hir::Either;
+use crate::db::RootDatabase;
+use test_utils::tested_by;
+
+pub enum NameRefKind {
+    Method(hir::Function),
+    Macro(hir::MacroByExampleDef),
+    FieldAccess(hir::StructField),
+    AssocItem(hir::ImplItem),
+    Def(hir::ModuleDef),
+    SelfType(hir::Ty),
+    Pat(AstPtr<ast::Pat>),
+    SelfParam(AstPtr<ast::SelfParam>),
+    GenericParam(u32),
+}
+
+pub(crate) fn classify_name_ref(
```

Comment:
I should probably collapse the `None` into `NameRefKind`.

Reply:
Hm, I think `Option<NameRefKind>` makes more sense, as it works with `?`